### PR TITLE
feat/Restart IR services after neo-go is down

### PR DESCRIPTION
**Pull request by notimetoname**
Relates #1337. IR only. Tested with https://github.com/notimetoname/neofs-dev-env/tree/local/add-rpc-morph-node.

### CODE REVIEW

File: pkg/morph/client/multi.go [hunk 1]
```diff
@@ -15,21 +15,28 @@ type Endpoint struct {
...
 		select {
 		case <-time.After(c.cfg.reconnectionDelay):
 		case <-c.closeChan:
+			c.switchLock.Unlock()
 			return false
 		}
 	}
 
 	c.inactive = true
+	c.switchLock.Unlock()
```
Comment (cthulhu-rider) [inline pkg/morph/client/multi.go:39]: placing this before statemenet below changes method behavior, i'd do it in a separate commit with explanation (i realize why u've done it). And then autonomous commit regarding the new callback

**Review by cthulhu-rider — COMMENTED** [review]
429e23a4734df9017fe529aeda09a3eacb23762d is useless to me, why don't we add this method along with the first use case?

**Comment by notimetoname** [comment]
> https://github.com/nspcc-dev/neofs-node/commit/429e23a4734df9017fe529aeda09a3eacb23762d is useless to me, why don't we add this method along with the first use case?

Different files, different structs (the first is about `innerRingIndexer`, the second about `Server`). But well, my args are not that strong, fixed them up.

File: pkg/innerring/innerring.go [hunk 2]
```diff
@@ -1011,8 +1015,9 @@ func createListener(ctx context.Context, cli *client.Client, p *chainParams) (ev
 	return listener, err
 }
 
-func createClient(ctx context.Context, p *chainParams, errChan chan<- error) (*client.Client, error) {
-	endpoints := p.cfg.GetStringSlice(p.name + ".endpoints")
+func (s *Server) createClient(ctx context.Context, p *chainParams, errChan chan<- error) (*client.Client, error) {
+	name := p.name
```
[the remark this answers is not in the document]
Comment (notimetoname) [reply pkg/innerring/innerring.go:1019]: > p is mutated on caller's side?

yes, spent some time fixing a bug that relates to the pointer. ok lets drop the pointer

File: pkg/innerring/innerring.go [hunk 3]
```diff
@@ -73,17 +73,19 @@ type (
 		blockTimers     []*timer.BlockTimer
 		epochTimer      *timer.BlockTimer
 
+		morphClient      *client.Client
```
[the remark this answers is not in the document]
Comment (notimetoname) [reply pkg/innerring/innerring.go:76]: the problem was: 90% of the issue fixing time i spend on code reading and realizing what is _the state_ and what is not _the state_. after the n-th time reading "// global state" block with the clients, i flipped out and decided to split them. so:
1. a separate commit OR
2. you are against that move at all?

File: pkg/innerring/innerring.go [hunk 4]
```diff
@@ -1145,3 +1163,30 @@ func (s *Server) newEpochTickHandlers() []newEpochHandler {
 
 	return newEpochHandlers
 }
+
+func (s *Server) restartMorph() error {
+	s.log.Info("restarting internals because of RPC connection loss...")
```
[the remark this answers is not in the document]
Comment (notimetoname) [reply pkg/innerring/innerring.go:1168]: > what exactly've been restarted

i would not do that. their number can grow up and i doubt a dev would fix that message

> or at least services|internal services

fixed

File: pkg/util/precision/converter.go [hunk 5]
```diff
@@ -22,7 +22,7 @@ type (
 	// This is a JSON bound that uses neo node. Neo-go has int64 limit for
 	// `smartcontract.Parameter` of integer type.
 	Fixed8Converter struct {
-		m *sync.RWMutex
+		m sync.RWMutex
```
[the remark this answers is not in the document]
Comment (notimetoname) [reply pkg/util/precision/converter.go:25]: no idea how it happened, it was a mistake, fixed

File: pkg/innerring/innerring.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cthulhu-rider) [reply pkg/innerring/innerring.go:76]: 1 ofc: the change itself can be really useful for readability/maintainability, especially since u stumble

**Review by cthulhu-rider — APPROVED** [review]

**Review by cthulhu-rider — APPROVED** [review]

File: pkg/util/precision/converter.go [hunk 6]
```diff
@@ -21,6 +22,7 @@ type (
 	// This is a JSON bound that uses neo node. Neo-go has int64 limit for
 	// `smartcontract.Parameter` of integer type.
 	Fixed8Converter struct {
+		m sync.RWMutex
```
Comment (roman-khimov) [inline pkg/util/precision/converter.go:25]: Is it really worth the trouble? Why not recreating the converter where needed?

File: pkg/util/precision/converter.go [hunk 7]
```diff
@@ -71,6 +79,9 @@ func (c *Fixed8Converter) SetBalancePrecision(precision uint32) {
 		exp = -exp
 	}
 
+	c.m.Lock()
+	defer c.m.Unlock()
```
Comment (roman-khimov) [inline pkg/util/precision/converter.go:83]: Frankly, I'd just internalize this `SetBalancePrecision` and leave it all to `NewConverter`.

File: pkg/innerring/innerring.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/innerring/innerring.go:76]: Option 1, yes, it looks suspicious now.

File: pkg/innerring/innerring.go [hunk 8]
```diff
@@ -1145,3 +1163,30 @@ func (s *Server) newEpochTickHandlers() []newEpochHandler {
 
 	return newEpochHandlers
 }
+
+func (s *Server) restartMorph() error {
```
Comment (roman-khimov) [inline pkg/innerring/innerring.go:1166]: I've been expecting some `innerring.New()` instead of this, but looking at the way connections are managed now maybe it's not realistic...

File: pkg/util/precision/converter.go (hunk 7 above)
Comment (notimetoname) [reply pkg/util/precision/converter.go:83]: dont you like how every blockchain config is updated [the same way](https://github.com/notimetoname/neofs-node/blob/92c391dda95e151bc96ca55d9242b20c8406d6f4/pkg/innerring/innerring.go#L1101-L1104) (i even have not created any func for that, `initConfigFromBlockchain` already exists!)?

File: pkg/innerring/innerring.go (hunk 8 above)
Comment (notimetoname) [reply pkg/innerring/innerring.go:1167]: so that is why issue fixing is so slow. i can change it but i dont know what way is the best (for now), so after trying some ways i came up with the shortest one (111 lines only and it works)

> I've been expecting some innerring.New()

do you mean the full [New](https://github.com/notimetoname/neofs-node/blob/92c391dda95e151bc96ca55d9242b20c8406d6f4/pkg/innerring/innerring.go#L329) every time a connections is lost?

also, left https://github.com/nspcc-dev/neofs-node/issues/2410

File: pkg/util/precision/converter.go (hunk 6 above)
Comment (roman-khimov) [reply pkg/util/precision/converter.go:25]: I mean using `NewConverter()` when needed (on reinit). Once initialized it never changes. If connection is broken then all users of it terminate anyway. Once it's up again we'll have a new converter and they will use this new one. Something like that (I'm not entirely sure it fits the current code well, but it theory it should).

File: pkg/util/precision/converter.go (hunk 7 above)
Comment (roman-khimov) [reply pkg/util/precision/converter.go:83]: Epoch-related things _can_ change without reconnecting, balance precision can't.

File: pkg/innerring/innerring.go (hunk 8 above)
Comment (roman-khimov) [reply pkg/innerring/innerring.go:1167]: The problem is that we have dependencies inverted here. There can't be any IR unless there is a chain (well, two chains) and there is a connection to it. But we have this IR struct that pretends to be independent of its own state (that is inherently tied to chains) and handles connections internally. It can accept current solution if refactoring it now requires too much effort.

File: pkg/util/precision/converter.go (hunk 6 above)
Comment (notimetoname) [reply pkg/util/precision/converter.go:25]: > Once initialized it never changes.

oh, you mean that we can not take into account that [precision](https://github.com/notimetoname/neofs-node/blob/0aee98732d75f67446df2167a294e560bad74039/pkg/util/precision/converter.go#L76) may change while we were down? or i got you wrong one more time?

File: pkg/util/precision/converter.go (hunk 7 above)
Comment (notimetoname) [reply pkg/util/precision/converter.go]: > balance precision can't

oh, really? it simplifies things a little.

File: pkg/util/precision/converter.go (hunk 6 above)
Comment (notimetoname) [reply pkg/util/precision/converter.go]: made the precision constant

File: pkg/innerring/innerring.go (hunk 8 above)
Comment (notimetoname) [reply pkg/innerring/innerring.go:1132]: > It can accept current solution if refactoring it now requires too much effort.

refactor will be done [separately](https://github.com/nspcc-dev/neofs-node/issues/2410#issuecomment-1613554057)

File: pkg/util/precision/converter.go (hunk 6 above)
Comment (roman-khimov) [reply pkg/util/precision/converter.go]: I take as [NEP-17 `decimals()`](https://github.com/neo-project/proposals/blob/06a6b550987a74025009cf6841757f2426462697/nep-17.mediawiki#decimals). It can't change for a contract. And it's the same here, no one should even of changing it on a real network. It can only happen if you're to create another contract, another token and migrate to it. But this is a completely different story.

**Review by roman-khimov — APPROVED** [review]
